instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpClientPeerServiceAttributesExtractor.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.http;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.net.PeerServiceResolver;
+import io.opentelemetry.instrumentation.api.instrumenter.url.UrlParser;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import javax.annotation.Nullable;
+
+/**
+ * Extractor of the {@code peer.service} span attribute, described in <a
+ * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes">the
+ * specification</a>.
+ */
+public final class HttpClientPeerServiceAttributesExtractor<REQUEST, RESPONSE>
+    implements AttributesExtractor<REQUEST, RESPONSE> {
+
+  private final HttpClientAttributesGetter<REQUEST, RESPONSE> attributesGetter;
+  private final PeerServiceResolver peerServiceResolver;
+
+  // visible for tests
+  HttpClientPeerServiceAttributesExtractor(
+      HttpClientAttributesGetter<REQUEST, RESPONSE> attributesGetter,
+      PeerServiceResolver peerServiceResolver) {
+    this.attributesGetter = attributesGetter;
+    this.peerServiceResolver = peerServiceResolver;
+  }
+
+  /**
+   * Returns a new {@link HttpClientPeerServiceAttributesExtractor} that will use the passed {@code
+   * attributesGetter} instance to determine the value of the {@code peer.service} attribute.
+   */
+  public static <REQUEST, RESPONSE>
+      HttpClientPeerServiceAttributesExtractor<REQUEST, RESPONSE> create(
+          HttpClientAttributesGetter<REQUEST, RESPONSE> attributesGetter,
+          PeerServiceResolver peerServiceResolver) {
+    return new HttpClientPeerServiceAttributesExtractor<>(attributesGetter, peerServiceResolver);
+  }
+
+  @Override
+  public void onStart(AttributesBuilder attributes, Context parentContext, REQUEST request) {}
+
+  @Override
+  public void onEnd(
+      AttributesBuilder attributes,
+      Context context,
+      REQUEST request,
+      @Nullable RESPONSE response,
+      @Nullable Throwable error) {
+
+    if (peerServiceResolver.isEmpty()) {
+      // optimization for common case
+      return;
+    }
+
+    String serverAddress = attributesGetter.getServerAddress(request);
+    Integer serverPort = attributesGetter.getServerPort(request);
+    String path = getUrlPath(attributesGetter, request);
```

Comment:
> it would be nice if we could avoid the url parsing in the common case where there's no peer service mapping (and even better also in the case where there's no peer service mapping that depends on path)

It would be nice indeed !

> what do you think about passing in the http.url (and only the http.url) to the PeerServiceResolver?

I don't mind seing the PeerServiceResolver having different interfaces depending on the client instrumentation type. Does it has an impact on the agent configuration like another property or having to precise the port explicitly ?

> PeerServiceResolver could then just check matches directly against the http.url without needing to do any parsing I think

That seems quite clean !

I see your point on performance here but I don't see yet what will be the impact on the configuration of the agent and how it would be implemented. 
If you want to proceed on that trajectory, feel free to take it from here.